deepspeed/utils/tensor_fragment.py

```diff
@@ -58,9 +61,56 @@ def get_full_hp_param(self, optim_state_key=None):
     return reduce_buffer.reshape_as(self)
 
 
+def safe_get_full_fp32_param(param):
+    if hasattr(param, '_hp_mapping'):
+        return param.get_full_hp_param()
+    return None
+
+
+def safe_get_full_optimizer_state(param, optim_state_key):
+    if hasattr(param, '_hp_mapping'):
+        return param.get_full_hp_param(optim_state_key)
+    return None
+
+
+def get_full_hp_grad(self):
+    reduce_buffer = torch.zeros_like(self, dtype=torch.float32).flatten()
+    if self._hp_mapping is not None:
+        lp_frag_address = self._hp_mapping.lp_fragment_address
+        reduce_fragment = torch.narrow(reduce_buffer,
+                                       0,
+                                       lp_frag_address.start,
+                                       lp_frag_address.numel)
+
+        hp_frag_address = self._hp_mapping.hp_fragment_address
+        hp_mapping = self._hp_mapping
+        # lp_grad_fragment = torch.narrow(hp_mapping.gradient_dict[hp_mapping.param_group_index],
+        #                                 0,
+        #                                hp_frag_address.start,
+        #                                hp_frag_address.numel)
+
```

Comment:
```suggestion
        from deepspeed.utils import logger
        if hp_mapping.param_group_index not in hp_mapping.gradient_dict or hp_mapping.gradient_dict[hp_mapping.param_group_index] is None:
            logger.warning("Gradients are only available immediately after backward and before egine step")
            return None
```

Reply:
not sure if we want to warn or assert here though. You know my stance on warnings...

If the user calling this function they must want the gradients, so it's a usage error if they call it at the wrong timing and thus perhaps it should be an assert?

Reply:
I agree that assert is appropriate here.